### ISSUE TITLE
[OP] Add sampling technique that allows stratified sampling

### ISSUE THREAD

**Issue by benfred** [NVIDIA/NVTabular#38]
<a href="https://github.com/rnyak"><img src="https://avatars1.githubusercontent.com/u/16246900?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [rnyak](https://github.com/rnyak)**
_Tuesday May 19, 2020 at 20:38 GMT_
_Originally opened as https://github.com/rapidsai/recsys/issues/149_

----

**Is your operator request related to a problem? Please describe.**

This method was used in the W&D Outbrain repo to generate validation set. 

**Describe alternatives you've considered**
The method used is Spark's `sampleBy` function:

```
sampleBy(col, fractions, seed=None)¶
Parameters: 
col – column that defines strata
fractions – sampling fraction for each stratum. If a stratum is not specified, we treat its fraction as zero.
seed – random seed
Returns: a new DataFrame that represents the stratified sample
```

**Describe the solution you'd like**
A clear and concise description of the operation you'd like to perform on the column.  Please include:
 - Type: Preprocessing
 - input column type(s): Dataframe  and a specific column used as input- [categorical, continuous].
 - output column type(s): [categorical, continuous]
 - Expected transformation of the data after application: Returns a stratified sample with or without replacement based on the fraction given on each stratum.

**Additional context**
Sklearn and pandas have the existing sampling methods, whereas cudf does not have such operation yet:

1. `sklearn.model_selection.train_test_split(*arrays, **options)` [source](https://github.com/scikit-learn/scikit-learn/blob/fd237278e/sklearn/model_selection/_split.py#L2029)

2. `https://github.com/scikit-learn/scikit-learn/blob/fd237278e/sklearn/model_selection/_split.py#L2029`[source](https://github.com/pandas-dev/pandas/blob/v1.0.3/pandas/core/generic.py#L4866-L5062)

**Comment by benfred** [comment]
<a href="https://github.com/rnyak"><img src="https://avatars1.githubusercontent.com/u/16246900?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [rnyak](https://github.com/rnyak)**
_Thursday May 28, 2020 at 20:18 GMT_

----

@jperez999 @benfred what's your assessment how long this will take to develop? It is low priority but if it is quick to do we can add it.

**Comment by benfred** [comment]
We've added this to the outbrains example by going
```python
# # Do a stratified split of the merged dataset into a training/validation dataset
merged['day_event'] = (merged['timestamp'] / 1000 / 60 / 60 / 24).astype(int)
random_state = cudf.Series(cupy.random.uniform(size=len(merged)))
valid_set, train_set = merged.scatter_by_map(((merged.day_event <= 10) & (random_state > 0.2)).astype(int)) 
train_set.to_parquet(OUTPUT_BUCKET_FOLDER+"train_gdf.parquet", compression=None)
valid_set.to_parquet(OUTPUT_BUCKET_FOLDER+"valid_gdf.parquet", compression=None)
```
